### PR TITLE
base-files: include TARGET_PROFILE in OPENWRT_BOARD

### PR DESCRIPTION
**Pull request by pprindeville**
The value for OPENWRT_BOARD in /etc/os-release omits TARGET_PROFILE
but really should include it.  Management software running on a
platform could use a foolproof way to figure out which image a
box should be reprovisioned with/upgraded to.

In $(VERSION_SED_SCRIPT), we add the new escape '%X' which is like
'%S' but appends /$(PROFILE_SANITIZED) if non-empty.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

### CODE REVIEW

**Comment by ynezz** [comment]
> Management software running on a platform could use a foolproof way to figure out which image a box should be reprovisioned with/upgraded to.

You can use following:

```shell
ubus call system board | jsonfilter -e @.model -e @.board_name
```

So I don't think, that we need to change stable content of `OPENWRT_BOARD` variable.

**Comment by pprindeville** [comment]
> > Management software running on a platform could use a foolproof way to figure out which image a box should be reprovisioned with/upgraded to.
> 
> You can use following:
> 
> ```shell
> ubus call system board | jsonfilter -e @.model -e @.board_name
> ```
> 
> So I don't think, that we need to change stable content of `OPENWRT_BOARD` variable.

There's still an argument to be made for having all the information in one place and not having to figure out how to pick it out of various disparate locations in varying formats.

What about adding an addition variable, like `OPENWRT_PROFILE` to `/etc/os-release` if you don't want to change the "stable"[but incomplete] contents" of the existing values?

Please reopen this for more discussion.

**Comment by pprindeville** [comment]
> So I don't think, that we need to change stable content of OPENWRT_BOARD variable.

@ynezz Latest version doesn't modify existing variables in `/etc/os-release`.

**Comment by pprindeville** [comment]
From @ynezz email, attaching it here for posterity:

> `ubus call system board | jsonfilter -e @.model -e @.board_name`

**Comment by aparcar** [comment]
Did you verify that this works? I might be wrong here but I somewhat remember that the current buildbot shares a rootfs folder between images of the same target/subtarget combination, which would results in misleading results for nightly builds.

The ubus call mentioned before should contain all you need. I'm assuming you want to use a script anyway, so that's an extra line if I'm not mistaken?

Lastly, this is a recurring topic which I've seen before, since `board_name` is not always really `profile`, especially in case of x86. I'd be happiest if running devices become "self aware" of what image (profile) it is running.

**Comment by pprindeville** [comment]
> Did you verify that this works? I might be wrong here but I somewhat remember that the current buildbot shares a rootfs folder between images of the same target/subtarget combination, which would results in misleading results for nightly builds.

I've only tried to build images for the Archer A7 v5.  I've not used `CONFIG_TARGET_MULTI_PROFILE=y` for instance.

How should I test for the corner cases?

I see this in `include/image.mk` but haven't yet tried to figure out what the implications are:

```
   440	ifdef IB
   441	  DEVICE_CHECK_PROFILE = $(filter $(1),DEVICE_$(PROFILE) $(PROFILE))
   442	else
   443	  DEVICE_CHECK_PROFILE = $(CONFIG_TARGET_$(if $(CONFIG_TARGET_MULTI_PROFILE),DEVICE_)$(call target_conf,$(BOARD)$(if $(SUBTARGET),_$(SUBTARGET)))_$(1))
   444	endif
...
   455	define Device/Check/Common
   456	  _PROFILE_SET = $$(strip $$(foreach profile,$$(PROFILES) DEVICE_$(1),$$(call DEVICE_CHECK_PROFILE,$$(profile))))
   457	  DEVICE_PACKAGES += $$(call extra_packages,$$(DEVICE_PACKAGES))
   458	  ifdef TARGET_PER_DEVICE_ROOTFS
   459	    $$(eval $$(call merge_packages,_PACKAGES,$$(DEVICE_PACKAGES) $$(call DEVICE_EXTRA_PACKAGES,$(1))))
   460	    ROOTFS_ID/$(1) := $$(if $$(_PROFILE_SET),$$(call mkfs_packages_id,$$(_PACKAGES)))
   461	    PACKAGES_$$(ROOTFS_ID/$(1)) := $$(_PACKAGES)
   462	  endif
   463	endef
```

> The ubus call mentioned before should contain all you need. I'm assuming you want to use a script anyway, so that's an extra line if I'm not mistaken?

No, I'm doing it in a manager daemon that's C-based.  So reading text files is better than running a bunch of external scripts or pipelines.

> Lastly, this is a recurring topic which I've seen before, since `board_name` is not always really `profile`, especially in case of x86. I'd be happiest if running devices become "self aware" of what image (profile) it is running.

Agreed.

Too bad `uname(2)` doesn't have more fields...

**Review by outbackdingo — APPROVED** [review]

File: include/version.mk [hunk 1]
```diff
@@ -56,6 +56,10 @@ VERSION_PRODUCT:=$(if $(VERSION_PRODUCT),$(VERSION_PRODUCT),Generic)
 VERSION_HWREV:=$(call qstrip,$(CONFIG_VERSION_HWREV))
 VERSION_HWREV:=$(if $(VERSION_HWREV),$(VERSION_HWREV),v0)
 
+# borrowed from include/target.mk and include/image.mk
```
Comment (aparcar) [inline include/version.mk:59]: I don't thinks this line is needed since `git blame` shows the missing info.

**Comment by outbackdingo** [comment]
> > > Management software running on a platform could use a foolproof way to figure out which image a box should be reprovisioned with/upgraded to.
> > 
> > 
> > You can use following:
> > ```shell
> > ubus call system board | jsonfilter -e @.model -e @.board_name
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So I don't think, that we need to change stable content of `OPENWRT_BOARD` variable.
> 
> There's still an argument to be made for having all the information in one place and not having to figure out how to pick it out of various disparate locations in varying formats.
> 
> What about adding an addition variable, like `OPENWRT_PROFILE` to `/etc/os-release` if you don't want to change the "stable"[but incomplete] contents" of the existing values?
> 
> Please reopen this for more discussion.

We agree is it much easier and safer programmatically to utilize it this way for data retrieval then the suggested execing of a shell call method

**Comment by outbackdingo** [comment]
> From @ynezz email, attaching it here for posterity:
> 
> > `ubus call system board | jsonfilter -e @.model -e @.board_name`



From @ynezz email, attaching it here for posterity:

ubus call system board | jsonfilter -e @.model -e @.board_name
Failed to connect to ubus
Failed to parse json data: unexpected end of data

File: include/version.mk (hunk 1 above)
Comment (pprindeville) [reply include/version.mk]: Done

**Comment by aparcar** [comment]
> ubus call system board | jsonfilter -e @.model -e @.board_name

```
root@79c1e172dfe5:/# ubus call system board | jsonfilter -e @.model -e @.board_name
LENOVO 20Q1S02M00
lenovo-20q1s02m00
```

**Comment by adschm** [comment]
As stated earlier already, this PR simply does not work.

base-files package is created per (sub)target, so including device-specific info there is not possible (beyond switch-case scripts).

**Comment by pprindeville** [comment]
> As stated earlier already, this PR simply does not work.
> 
> base-files package is created per (sub)target, so including device-specific info there is not possible (beyond switch-case scripts).

@adschm What about putting the file into a variant sub-package?

**Comment by adschm** [comment]
> > As stated earlier already, this PR simply does not work.
> > base-files package is created per (sub)target, so including device-specific info there is not possible (beyond switch-case scripts).
> 
> @adschm What about putting the file into a variant sub-package?

There is simply no way to achieve what you want with a package.

Apart from that, I don't think that it's really desirable or necessary either.

**Comment by pprindeville** [comment]
> > > As stated earlier already, this PR simply does not work.
> > > base-files package is created per (sub)target, so including device-specific info there is not possible (beyond switch-case scripts).
> > 
> > 
> > @adschm What about putting the file into a variant sub-package?
> 
> There is simply no way to achieve what you want with a package.
> 
> Apart from that, I don't think that it's really desirable or necessary either.

We could modify `/etc/os-release` on firstboot... and append the necessary line.

If KISS is still a basic precept of good software design, why wouldn't we want to have all of the identifying information in one simple, easy to access place?

**Comment by pprindeville** [comment]
> Did you verify that this works? I might be wrong here but I somewhat remember that the current buildbot shares a rootfs folder between images of the same target/subtarget combination, which would results in misleading results for nightly builds.
> 
> The ubus call mentioned before should contain all you need. I'm assuming you want to use a script anyway, so that's an extra line if I'm not mistaken?
> 
> Lastly, this is a recurring topic which I've seen before, since `board_name` is not always really `profile`, especially in case of x86. I'd be happiest if running devices become "self aware" of what image (profile) it is running.

Point taken on `board_name` not being the same.

I guess what I really want (at least for now) is a way to locate __this__ image running on this box again, so I can retrieve an updated version of it.

The problem is that you don't have the full description of an image *within* an image.  Which is odd.  The outside of the image has a distinct name, but there's not enough information inside the image to figure out which one ended up being provisioned on the current box.

How do we fix this shortcoming?

**Comment by adschm** [comment]
> We could modify /etc/os-release on firstboot... and append the necessary line.
>
> If KISS is still a basic precept of good software design, why wouldn't we want to have all of the identifying information in one simple, easy to access place?

Well, os-release contains information about the release and it's configuration. This has nothing to do with the device, so the device information does not belong there. (Apart from the fact that I don't think editing this file via two completely different mechanisms and at two different times really makes the whole process better.)

**Comment by adschm** [comment]
> I guess what I really want (at least for now) is a way to locate **this** image running on this box again, so I can retrieve an updated version of it.
> 
> The problem is that you don't have the full description of an image _within_ an image. Which is odd. The outside of the image has a distinct name, but there's not enough information inside the image to figure out which one ended up being provisioned on the current box.

I think this is the nucleus of the misunderstanding. You don't want to get the exact image you installed with in most cases, as that would be a factory image that won't help you with OpenWrt installed. What you want is the correct image for your device, and that's what SUPPORTED_DEVICES and board_name are for.

**Comment by pprindeville** [comment]
> Well, os-release contains information about the release and it's configuration. This has nothing to do with the device, so the device information does not belong there. (Apart from the fact that I don't think editing this file via two completely different mechanisms and at two different times really makes the whole process better.)

Does "configuration" include the type of hardware it's running on?  It would seem to:

```
NAME="OpenWrt"
VERSION="1.3"
ID="openwrt"
ID_LIKE="lede openwrt"
PRETTY_NAME="OpenWrt 1.3"
VERSION_ID="1.3"
HOME_URL="https://openwrt.org/"
BUG_URL="https://bugs.openwrt.org/"
SUPPORT_URL="https://forum.openwrt.org/"
BUILD_ID="r0+16586-c7ab77ccc7"
OPENWRT_BOARD="ath79/generic"
OPENWRT_ARCH="mips_24kc"
OPENWRT_TAINTS="no-all busybox"
OPENWRT_DEVICE_MANUFACTURER="OpenWrt"
OPENWRT_DEVICE_MANUFACTURER_URL="https://openwrt.org/"
OPENWRT_DEVICE_PRODUCT="Generic"
OPENWRT_DEVICE_REVISION="v0"
OPENWRT_RELEASE="OpenWrt 1.3 r0+16586-c7ab77ccc7"
```

We've got all sorts of details about the hardware (architecture, target, and subtarget).  Why is the model of board it's running on a bridge-too-far?  It seems like an arbitrary distinction.

**Comment by pprindeville** [comment]
> I think this is the nucleus of the misunderstanding. You don't want to get the exact image you installed with in most cases, as that would be a factory image that won't help you with OpenWrt installed. What you want is the correct image for your device, and that's what SUPPORTED_DEVICES and board_name are for.

What if my definition of "the correct image" is a more recent build of the same architecture/target/subtarget/profile?

**Comment by adschm** [comment]
> > Well, os-release contains information about the release and it's configuration. This has nothing to do with the device, so the device information does not belong there. (Apart from the fact that I don't think editing this file via two completely different mechanisms and at two different times really makes the whole process better.)
> 
> Does "configuration" include the type of hardware it's running on? It would seem to:
> 
> [...]
> 
> We've got all sorts of details about the hardware (architecture, target, and subtarget). Why is the model of board it's running on a bridge-too-far? It seems like an arbitrary distinction.

Target/subtarget actually change the OpenWrt configuration (i.e. change the way packages are compiled and their content).
Devices in a subtarget then only change which packages are _selected_, and how the final image is put together. (That's BTW why your approach here is not working.) Thus, having target/subtarget inside os-release, but not having device/profile there actually makes sense on a theoretical basis, too.

**Comment by adschm** [comment]
> > I think this is the nucleus of the misunderstanding. You don't want to get the exact image you installed with in most cases, as that would be a factory image that won't help you with OpenWrt installed. What you want is the correct image for your device, and that's what SUPPORTED_DEVICES and board_name are for.
> 
> What if my definition of "the correct image" is a more recent build of the same architecture/target/subtarget/profile?
> 
> Also, there are a lot of platforms (x86 and x86_64 especially) for which the value in `/etc/board.json` doesn't match any value found in `SUPPORTED_DEVICES` anywhere.
> 
> Examples:
> 
> My Supermicro SYS-D5018-FN4T contains this:
> 
> ```
> 	"model": {
> 		"id": "supermicro-sys-5018d-fn8t",
> 		"name": "Supermicro SYS-5018D-FN8T"
> 	},
> ```
> 
> My PC Engines APU4 has:
> 
> ```
> 	"model": {
> 		"id": "pc-engines-pc-engines-apu4",
> 		"name": "PC Engines PC Engines apu4"
> 	},
> ```
> 
> and a KVM has:
> 
> ```
> 	"model": {
> 		"id": "red-hat-kvm",
> 		"name": "Red Hat KVM"
> 	},
> ```
> 
> none of which match `SUPPORTED_DEVICES` (because none are defined for `x86`).
> 

Having looked at this, I'm getting the strong feeling it's simply our x86 target not being aligned with our general concept of handling sysupgrade. Thinking of that, I remember that back when I put the global SUPPORTED_DEVICES variable into place, x86 already broke once because sysupgrade was not implemented correctly there [1].

Since we have the json metadata now that contains the supported_devices strings for all images, essentially the combination of board_name and supported_devices should allow to identify the matching image in the usual scenario. If x86 puts some specific strings there, but only provides unspecific images, this looks like a misimplementation* in x86 to me.

[1] https://github.com/openwrt/openwrt/commit/29167cbca3653de05a8b915bc21327dac7d05174

* does this word actually exist?

**Comment by pprindeville** [comment]
> Having looked at this, I'm getting the strong feeling it's simply our x86 target not being aligned with our general concept of handling sysupgrade. Thinking of that, I remember that back when I put the global SUPPORTED_DEVICES variable into place, x86 already broke once because sysupgrade was not implemented correctly there [1].

The x86 has long been a black sheep of OpenWRT, it's true, despite its obvious advantage to being able to easily boot it and test it as VM's in a CI/CD environment.

@dangowrt and @uxgood  Did great great work in getting `sysupgrade` running on it.  Why it continues to lag, I'm not sure.

Who doesn't have an old, underpowered PC lying around that they could easily throw a 2-port PCI-e or PCI Ethernet NIC into?

> Since we have the json metadata now that contains the supported_devices strings for all images, essentially the combination of board_name and supported_devices should allow to identify the matching image in the usual scenario. If x86 puts some specific strings there, but only provides unspecific images, this looks like a misimplementation* in x86 to me.

To be usable, the json metadata would need to be *inside the image* so an agent on the router can figure out what new image to fetch from a repo, which is what I've been asking for all along.

But I'm not sure I agree with that last sentence.  x86 has plug-n-play and BIOS to help in device discovery.  Which means you can have generic image classes that can run on a wide range of devices...

> [1] [29167cb](https://github.com/openwrt/openwrt/commit/29167cbca3653de05a8b915bc21327dac7d05174)
> 
> `*` does this word actually exist?

No, but we can coin it.

**Comment by aparcar** [comment]
Can't we just attach openwrt_profile=foobar to the kernel cmdline and extract that information later?

**Comment by pprindeville** [comment]
> Can't we just attach openwrt_profile=foobar to the kernel cmdline and extract that information later?

That would work.  It's not as nice as having everything in one place, but...

**Comment by aparcar** [comment]
On second thought this isn't super nice, a better approach could be to add `openrwt_profile` within DTS files next to `compatible` and `model`. I'll try to figure out a generic way.

Ultimately this would end up somewhere in `/proc/` and could be accessible via `ubus call system board`. For your convenience, something like `image_uri=$target/$subtarget/$profile` could be stored in `/tmp/sysinfo/image_uri`.

**Comment by pprindeville** [comment]
> On second thought this isn't super nice, a better approach could be to add `openrwt_profile` within DTS files next to `compatible` and `model`. I'll try to figure out a generic way.
> 
> Ultimately this would end up somewhere in `/proc/` and could be accessible via `ubus call system board`. For your convenience, something like `image_uri=$target/$subtarget/$profile` could be stored in `/tmp/sysinfo/image_uri`.

As I said to @adschm, the KISS approach is putting everything into `/usr/lib/os-release`.  Yes, this means touching this file both inside `base-files` as well as in `include/image.mk` but that shouldn't be the deciding factor.

**Comment by aparcar** [comment]
Looking at the code again I might have been wrong all along. The [buildbot uses](https://downloads.openwrt.org/snapshots/targets/mediatek/mt7622/config.buildinfo) the `CONFIG_TARGET_PER_DEVICE_ROOTFS` option, therefore the images would contain the desired information. It's just locally disabled per default, which should be possible to change. I'll do some compile testing and will give an update.

**Comment by aparcar** [comment]
The "per device rootfs" is a false hope since OpenWrt optimizes filesystems per package combination, see picture below. These images are simply recycled between image builds if the same package selection is used, meaning we can't write a file in them. I don't want to remove that optimization.
image

Fun other options:
* Modify the kernel CMDLINE to contain something like `openwrt_profile=<foobar>`
* Modify the individual DTS files to contain a `openwrt_profile` variable which would end up in `/proc/firmware/`
* Label the filesystems within the `append-rootfs` step on a per image basis (suggested by @dangowrt )

**Comment by pprindeville** [comment]
One other "fun option" might be to modify `/usr/lib/os-release` on first boot with the correct additional information.

But I don't see why we're enslaved to not breaking the optimizations you allude to.  If it's building us what we **need**, then who cares how optimized it is?  I'd rather have the needed results slowly, than have incomplete results quickly.

**Comment by adschm** [comment]
> On second thought this isn't super nice, a better approach could be to add `openrwt_profile` within DTS files next to `compatible` and `model`. I'll try to figure out a generic way.
> 
> Ultimately this would end up somewhere in `/proc/` and could be accessible via `ubus call system board`. For your convenience, something like `image_uri=$target/$subtarget/$profile` could be stored in `/tmp/sysinfo/image_uri`.

NAK. Please don't bloat the DTS files which additional variables that just do about the same as the already existing model and compatible.

**Comment by adschm** [comment]
> Looking at the code again I might have been wrong all along. The buildbot uses the CONFIG_TARGET_PER_DEVICE_ROOTFS option, therefore the images would contain the desired information. It's just locally disabled per default, which should be possible to change. I'll do some compile testing and will give an update.

This is a nice example why all build-time solutions are prone to fail: If somebody builds all devices without the mentioned option, any profile-related variable will be wrong for sure.

Anyway, this approach is conceptually mistaken as I tried to explain several times already, so I'm ending this now.

**Comment by outbackdingo** [comment]
all solved on our end with, which works!
get_os(){ ubus call system board | grep description | awk -F\" '{ print $4 }'; }
get_system(){ ubus call system board | grep system | awk -F\" '{ print $4 }'; }
get_model(){ cat /tmp/sysinfo/model; }

**Comment by pprindeville** [comment]
That generates meaningless output in the case of x86 hardware.

**Comment by outbackdingo** [comment]
okay pllease explain to me how this is meaningless data on x86, its exactly what is needed!

root@52-54-00-F9-B5-DD:~# cat /tmp/sysinfo/model
QEMU Standard PC (Q35 + ICH9, 2009)
root@52-54-00-F9-B5-DD:~# ubus call system board | grep description | awk -F\" '{ print $4 }'
OpenWrt SNAPSHOT r16835-f755d81999
root@52-54-00-F9-B5-DD:~# ubus call system board | grep system | awk -F\" '{ print $4 }'
Intel Xeon Processor (Cascadelake)

**Comment by pprindeville** [comment]
> okay pllease explain to me how this is meaningless data on x86, its exactly what is needed!
> 
> QEMU Standard PC (Q35 + ICH9, 2009)
> 
> OpenWrt SNAPSHOT r16835-f755d81999
> 
> Intel Xeon Processor (Cascadelake)

An x86 or x86_64 can either require BIOS or EFI boot.  And it can either map to an existing canned image with baked in platform drivers (as is the case of the Geos or PC Engines Alix and APU platforms) or it can take a generic PC image (either x86_32 or x86_64)... which an explicit profile would go a long way to identifying.
